### PR TITLE
MONGOCRYPT-482 Support BSON double for range algorithm

### PR DESCRIPTION
**Pull request by kevinAlbs**
# Summary

- Support BSON double for range insert and find.
- Remove unnecessary call to `mc_getTypeInfo*` in `get_edges`.

# Background & Summary

Test data was obtained by printing output from server unit tests:
https://github.com/kevinAlbs/mongo/blob/range_support/kevinAlbs/output.txt

[SERVER-70066](https://jira.mongodb.org/browse/SERVER-70066) prohibited index min/max being specified on double. Therefore, this PR does not include min/max options in `mc_getTypeInfoDouble_args_t`, `mc_getEdgesDouble_args_t`, or `mc_getMincoverDouble_args_t`. The index min/max value in the FLE2FindRangeSpec is still used to determine the default upper or lower bound. This matches [server behavior](https://github.com/mongodb/mongo/blob/master/src/mongo/crypto/fle_crypto.cpp#L1499-L1508).

### CODE REVIEW

File: edges_double.cstruct [hunk 1]
```diff
@@ -0,0 +1,6596 @@
+        {&getEdgesDouble, 0.00000000000000000, 0.00000000000000000, 12.33999999999999986, 1, {
```
Comment (eramongodb) [inline edges_double.cstruct:1]: Stray file?

Comment (kevinAlbs) [reply edges_double.cstruct]: Yes. Removed.

File: src/mongocrypt-marking.c [hunk 2]
```diff
@@ -606,48 +606,40 @@ get_edges (mc_FLE2RangeInsertSpec_t *insertSpec,
...
-      ((void) (0))
 
    if (value_type == BSON_TYPE_INT32) {
-      GET_AND_RETURN_EDGES (32, int32);
+      mc_getEdgesInt32_args_t args = {
+         .value = bson_iter_int32 (&insertSpec->v),
+         .min = OPT_I32 (bson_iter_int32 (&insertSpec->min)),
+         .max = OPT_I32 (bson_iter_int32 (&insertSpec->max)),
+         .sparsity = sparsity};
+      return mc_getEdgesInt32 (args, status);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply src/mongocrypt-marking.c]: Done. Thank you for the suggested changes.

File: test/test-mc-range-edge-generation.c [hunk 3]
```diff
@@ -197,6 +199,72 @@ _test_getEdgesInt64 (_mongocrypt_tester_t *tester)
...
+         ASSERT_STATUS_CONTAINS (status, test->expectError);
+         mongocrypt_status_destroy (status);
+         continue;
+      }
+      ASSERT_OK_STATUS (got != NULL, status);
+
+      size_t numGot = mc_edges_len (got);
+      size_t numExpected = 0;
+      while (test->expectEdges[numExpected] != NULL) {
+         numExpected += 1;
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-edge-generation.c]: Done. Applied to other occurrences in this file.

File: test/test-mc-range-encoding.c [hunk 4]
```diff
@@ -270,11 +272,95 @@ _test_RangeTest_Encode_Int64 (_mongocrypt_tester_t *tester)
...
+      /* Test cases copied from server Double_Errors test ... end */
+   };
+
+   for (size_t i = 0; i < sizeof (tests) / sizeof (tests[0]); i++) {
+      DoubleTest *test = tests + i;
+      mongocrypt_status_t *const status = mongocrypt_status_new ();
+
+      printf ("_test_RangeTest_Encode_Double: value=%f\n", test->value);
+      mc_OSTType_Double got;
+      bool ok = mc_getTypeInfoDouble (
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-encoding.c]: Done. Applied to other occurrences in this file.

File: test/test-mc-range-encoding.c [hunk 5]
```diff
@@ -270,11 +272,95 @@ _test_RangeTest_Encode_Int64 (_mongocrypt_tester_t *tester)
...
+         ASSERT_OR_PRINT_MSG (!ok, "expected error, but got none");
+         ASSERT_STATUS_CONTAINS (status, test->expectError);
+         mongocrypt_status_destroy (status);
+         continue;
+      }
+      ASSERT_OK_STATUS (ok, status);
+      ASSERT_CMPUINT64 (got.value, ==, test->expect);
+      ASSERT_CMPUINT64 (got.min, ==, 0);
+      ASSERT_CMPUINT64 (got.max, ==, UINT64_MAX);
+      mongocrypt_status_destroy (status);
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-encoding.c]: Done. Applied to other occurrences in this file.

File: test/test-mc-range-mincover.c [hunk 6]
```diff
@@ -243,6 +296,29 @@ _test_dump_64 (void *tests, size_t idx, mc_mincover_t *got)
...
+{
+   BSON_ASSERT_PARAM (tests);
+   DoubleTest *const test = (DoubleTest *) tests + idx;
+   fflush (stdout); // Avoid incomplete stdout output from prior tests on error
+   fprintf (stderr,
+            "testcase: lowerBound=%f %s upperBound=%f %s",
+            test->lowerBound,
+            test->includeLowerBound ? "included" : "excluded",
+            test->upperBound,
+            test->includeUpperBound ? "included" : "excluded");
```
[the remark this answers is not in the document]
Comment (kevinAlbs) [reply test/test-mc-range-mincover.c]: Done. Applied to other occurrences in this file.

**Review by eramongodb — APPROVED** [review]

**Review by vector-of-bool — APPROVED** [review]

**Pull request merged**